static_frame/core/index_hierarchy.py

```diff
@@ -1018,259 +1362,580 @@ def values_at_depth(self,
             sel = depth_level
         else:
             sel = list(depth_level)
+
         return self._blocks._extract_array(column_key=sel)
 
+    # NOTE: This is much slower than old impl. Not sure how to optimize.
     @doc_inject()
-    def label_widths_at_depth(self,
+    def label_widths_at_depth(self: IH,
             depth_level: DepthLevelSpecifier = 0
             ) -> tp.Iterator[tp.Tuple[tp.Hashable, int]]:
-        '''{}'''
-        if isinstance(depth_level, INT_TYPES):
-            sel = depth_level
-        else:
-            raise NotImplementedError('selection from iterables is not implemented')
-        yield from self._levels.label_widths_at_depth(depth_level=depth_level)
+        '''
+        {}
+        '''
+        pos: tp.Optional[int] = None
+
+        if depth_level is None:
+            raise NotImplementedError('depth_level of None is not supported')
+
+        if not isinstance(depth_level, INT_TYPES):
+            sel = list(depth_level)
+            if len(sel) == 1:
+                pos = sel.pop()
+        else: # is an int
+            pos = depth_level
+
+        if pos is None:
+            raise NotImplementedError(
+                'selecting multiple depth levels is not yet implemented'
+            )
+
+        if self._recache:
+            self._update_array_cache()
+
+        # Could this be memoized?
+        def _extract_counts(
+                arr: np.ndarray,
+                pos: int,
+                ) -> tp.Iterator[tp.Tuple[tp.Hashable, int]]:
+            unique, widths = ufunc_unique1d_counts(arr)
+            labels = self._indices[pos].values[unique]
+            yield from zip(labels, widths)
```

Comment:
Should this just be a class or module-level function?

Reply:
It needs `self._indices`!